Edge-based Contraction Hierarchies and Turncost support

easbar opened the pull request:
This PR shall enable turn costs for contraction hierarchies as described in this [paper](https://pdfs.semanticscholar.org/4aa9/65d323ccb8b9e7c71de77f910900ff69a773.pdf) by R. Geisberger and C.Vetter.

This PR fixes #270 and has the same goal as #912, but follows a rather different approach and is much closer to what is described in the paper. Several of the problems described in #912 are addressed here like multiple loops/p-turns, u-turn detection and fixed size turn cost tables. Fundamentally this is realized by storing two additional integers per shortcut edge, which are set to the
IDs of the first and last original edges that are skipped by the shortcut. Doing this requires
many adjustments of the CH query algorithm (when expanding nodes and when checking bridge nodes where forward and backward searches meet) as well as the node contraction / shortcut searches.

### What is working so far:

* (Hopefully) error-free implementation of the node contraction and query algorithms. There are extensive tests that try to cover all edge cases, some of them compare the query results with the standard Dijkstra implementation for many queries and node contraction orders. So far it looks like all situations are handled correctly. The U-Turn based traversal mode is not yet supported and requires further changes and tests (if it should be supported), see comment below. 

### Next steps:

* Integration with previous (node-based) CH code (`NodeContractor` and `PrepareContractionHierarchies`) and support for automatic node ordering. So far the nodes can only be ordered in random order or ordered by ID, but this makes no sense for larger graphs. So far `EdgeBasedNodeContractor` does not return any metrics that can be used for node ordering heuristics.

* Integration with Graphhopper API and integration tests, see what the performance is like.

* Performance optimizations: There are many obvious things that can be improved:

   - Probably the most important thing is reusing data structures between the different witness path searches, something like `DijkstraOneToMany`. For the turn replacement difference algorithm that is being used we need a way to initialize the search with edges adjacent to the start node and it must be possible to initialize them with arbitrary weights.

   - Tuning of the node contraction parameters. The above mentioned paper describes a different metric than the one that is currently used for example. We can start with something simple (edge difference / quotient only for example) and see what the performance is like.

   - Minor stuff: See some of the todos in the code, there is definitely room for improvements.

ammagamma commented:
Looks like `CHTurnCostTest#testFindPath_highlyConnectedGraph_compareWithDijkstra` still fails sometimes. This one is the ultimate bug detector ;)

ammagamma commented:
Regarding u-turn based traversal mode: For edge based CH we have to remember in quite a few places to exclude u-turns, which makes it hard sometimes. Would it not be much easier to set the default u-turn cost in TurnWeighting to infinity instead of explicitly checking if we make a u-turn (something like `if (!traversalMode.hasUTurnSupport() && edgeId == previousId)`) ? Maybe I am missing something, but it looks like we would not need an extra traversal mode if we would only adjust the default u-turn costs ? Or is it done this way because of performance reasons ?

boldtrn commented:
Great PR! Thanks for this.

> if (!traversalMode.hasUTurnSupport() && edgeId == previousId)

I just recently worked on recognizing u-turns. Unfortunately it is not as easy as checking if edgeId == previousId, this should only happen when you place a via point, but each path between two points is calculated separately so this case should not really occur for route calculation itself. More frequently you would face situations that are for example double left or double right turns resulting on the opposite lane of a lane separated street. These are different edges in GH and different ways in OSM, so the id would not be equal.

For example [this](https://graphhopper.com/maps/?point=48.643033%2C9.331577&point=48.643012%2C9.331759&locale=en-AU&vehicle=car&weighting=fastest&elevation=true&use_miles=false&layer=Omniscale) U-turn won't be recognized by this technique.

The recognizing of these u-turns currently only happens in the instruction calculation, see [here](https://github.com/graphhopper/graphhopper/blob/d1d60828ce4493c799990b5a817272d101406af4/core/src/main/java/com/graphhopper/routing/InstructionsFromEdges.java#L212).

karussell commented:
Cool, thanks a lot for this!

> Or is it done this way because of performance reasons ?

The idea was to allow u-turns at certain places to avoid large detours because of some turn restrictions, but as they are implemented they are too slow and also the routes feel often wrong as there might be u-turns at any junction. We could limit those junctions to e.g. dead ends but this could be done with the EDGE_BASED_2DIR traversal too IMO but still the u-turn costs would not be applied to situations mentioned from @boldtrn.

karussell commented on core/src/main/java/com/graphhopper/routing/GenericDijkstraBidirection.java:
```diff
@@ -0,0 +1,212 @@
+ *
+ * @author Peter Karich
+ */
+public abstract class GenericDijkstraBidirection<T extends SPTEntry> extends AbstractBidirAlgo {
```
Nice, probably we could merge with AbstractBidirAlgo and use GenericDijkstraBidirection for bidirectional A* too then.

ammagamma replied:
Yes I thought about this too. There should be quite a bit of code in Astar that can be reused here, but I have not looked close yet.

karussell commented on core/src/main/java/com/graphhopper/routing/DijkstraBidirectionEdgeCHNoSOD.java:
```diff
@@ -0,0 +1,221 @@
+        }
+        // we can not use in/outEdgeExplorer here because this method will be called within a loop using the 
+        // in/outEdgeExplorer already
+        EdgeExplorer edgeExplorer = graph.createEdgeExplorer(new DefaultEdgeFilter(flagEncoder, reverse, !reverse));
```
As creating EdgeExplorers is very expensive we should probably create two additional explorers in the constructor.

ammagamma replied:
Sure.

ammagamma commented:
@boldtrn @karussell 
> Unfortunately it is not as easy as checking if edgeId == previousId

Yes. Another possibiility where this also fails is if two edges with fwd flags in different directions are used instead of one edge with fwd&bwd, but the osm import does not do this so far I think. In this PR the goal will first be to get the same results as Dijkstra with EDGE_BASED_2DIR and when this works we could try to do the same with EDGE_BASED_2DIR_UTURN. Other kind of u-turns (for example lane switches like you mentioned) that are not detected by the current Dijkstra implementation should probably be first fixed with Dijkstra before we implement this for CH. What do you think ?

> The idea was to allow u-turns at certain places

Could we not use a default u-turn cost of infinity and use some finite value at the junctions where it should be allowed ? In the algorithms we would then stop expanding nodes/edges when the weight is infinity. In this case the u-turn behavior could be decided via the turn cost tables and the turn weightings and the algorithms would stay agnostic of the desired u-turn behavior. Just an idea, not sure...

karussell commented:
> What do you think ?

yes, this is for another issue

> Just an idea, not sure...

Yes, sounds good. Then we could reduce the traversal modes to just the two node vs. edge based.

ammagamma commented:
### Update:

- Integration is mostly done. We can now use the GraphHopper API to prepare a given map for edge-based CH and run CH routing queries with turn costs, just use for example `graph.flag_encoders=car|turn_costs=true` and you no longer need to set `prepare.ch.weightings=no`. Here is an example from the integration test in Moscow:
![image](https://user-images.githubusercontent.com/17603532/34425510-c7f36758-ec2c-11e7-80b2-cbf1a7e23f9c.png)
- The node contraction order is now determined automatically (using only the number of potential shortcuts as heuristic at the moment). Using a common `NodeContractor` interface with different implementations for edge/node-based graph traversal we can re-use the contraction ordering logic from `PrepareContractionHierarchies`.
- The results of the witness path searches are now reused for different outgoing edges in WitnessPathFinder. 
- The witness path searches are now limited by a maximum number of settled edges (currently hard-coded). This is more complicated than in the node-based case, because the witness path searches are not only needed to prove that some shortcut is not needed, but also to find the shortest path between two nodes (including loops at the node to be contracted).

The biggest issue right now is performance, for example the preparation for Germany's smallest state Bremen takes around `45s` and the average query time is about `5ms`. There are certainly many small/medium possible improvements, many of which are written as todos in the code, but I think first we should try to reduce the number of shortcuts that get introduced. For example at the moment for the Berlin map there are ~100 million shortcuts compared to ~2 million shortcuts in the node-based case without turn costs. This should reduce the performance of both the preparation and the routing queries. It is probably reasonable that in the edge-based case more shortcuts are required, but hopefully not this many. Right now I am optimistic that it can be improved, for example I have not tried the node priority metric given in the paper yet.

karussell commented:
Have not looked deeply in the sources but tried it and works nicely! (I am using this map https://ahorn.lima-city.de/tr/ to easily see OSM turn restrictions)

I can confirm that many more shortcuts are created (for my berlin case it was just a factor of 8) but the preparation took much longer: 40min vs. 8 seconds

> just use for example graph.flag_encoders=car|turn_costs=true and you no longer need to set prepare.ch.weightings=no

The config graph.flag_encoders=car|turn_costs=true is already usable for master too, but it is ignored for CH, so no need to set weightings to `no`. BTW: this makes a new parameter like prepare.ch.turn_costs=true for this feature necessary (when merged), until we improved the performance of preparation, shortcut number and query speed to an acceptable degree.

Most important message: it is really nice that now we have a working implementation that 'just' needs a bit tuning - this was a long way, thanks a lot :) !

karussell commented:
> For example at the moment for the Berlin map there are ~100 million shortcuts compared to ~2 million shortcuts in the node-based case without turn costs.

Ah, strange. Are you including brandenburg? For [just Berlin](http://download.geofabrik.de/europe/germany/berlin.html) I get ~500K shortcuts compared to 84K for master. The edge count is 112K and 79K nodes

ammagamma commented:
Thanks for testing! 

> Most important message: it is really nice that now we have a working implementation that 'just' needs a bit tuning 

Yes, so far I mostly focused on 'working' and 'tested', 'fast' and 'efficient' should be next ;)

> I am using this map https://ahorn.lima-city.de/tr/ to easily see OSM turn restrictions
Ah thats nice, do you use it just to find junctions with turn restrictions or can we also use it as map layer and show our routes ?

>  For just Berlin I get ~500K shortcuts compared to 84K for master. The edge count is 112K and 79K nodes. 

Yes, your numbers are correct, I was using the same map, but accidentally was looking at the number of dijkstras not shortcuts in the log. So for Berlin there are about six (not fifty, like I wrote) times more shortcuts (good news ;)

> this makes a new parameter like prepare.ch.turn_costs=true for this feature necessary

Should we be able to switch on/off turn costs at query time provided that the preparation has been done for node & edge based cases ? See `GraphHopperIT#testTurnCostsOnOffCH` ? For the request this can be done using the `Routing.EDGE_BASED` hint in `GHRequest`, but I am not sure how we should support both preparations. For example at the moment we cannot add multiple `CarFlagEncoder`, [something like `new EncodingManager("car,car|turn_costs=true")`], because they all have the same name "car".
Why do you think do we need a new parameter `prepare.ch.turn_costs=true` ? Right now I am enabling turn costs via the flag_encoder property (`graph.flag_encoders=car|turn_costs=true`). Using the new property we would still be restricted to have this setting globally for all encoders. Not sure if we need something like CH with turn costs for cars and without turn costs for bikes etc.

karussell commented:
> Should we be able to switch on/off turn costs at query time provided that the preparation has been done for node & edge based cases ?

Here I was just thinking about how early we merge this PR. If we merge it early to avoid too many update work we need the old "node based way" for CH even if there is turn_costs=true specified (which would be then only used for the non-CH stuff like it is done at the moment). If we merge when query & preparation performance is close or just 2x away no backward compatibility config is necessary.

ammagamma commented:
> I was just thinking about how early we merge this PR
Ok, right now I do not see a big problem having this on a separate branch and I would do the / some  performance improvements before merging. I could also pick some changes that are not only related to turn costs from this branch and put them in a separate PR.

ammagamma commented:
**Update**: I have made some improvements. Besides some minor improvements local searches are now cancelled entirely as soon as it is clear that the target edge can not be reached via the node to be contracted, it is checked if a shortcut that is about to be added exists already and I implemented the node contraction priority calculation as given in the paper.

For [Berlin](http://download.geofabrik.de/europe/germany/berlin.html) the contraction now takes about 4 minutes on my machine (about 15min previously) and I measured an average query time of about 13ms. For [Bremen](http://download.geofabrik.de/europe/germany/bremen.html)) the contraction takes about 25s contraction time and the queries run about 3ms.

However, the number of shortcuts is still very high and I am struggling to find a way to reduce it. For Berlin I am getting about 430k and for Bremen about 110k. The high density of shortcuts edges seems to limit the performance of both the graph contraction and the routing queries. Especially towards the end of the contraction there seem to be just too many shortcut edges that get introduced. Right now I am still hoping to find an error that is responsible for the large number of shortcuts, but having checked many things I am slowly running out of ideas what could be wrong.

Here is a list of things I have already looked into so far (but did not help much at least when I tried them): 

1.  Use an array- (instead of map-)based version for WitnessPathFinder similar to DijkstraOneToMany -> not too much improvement and does not help with the number of shortcuts (currently not commited in this branch)

2. node priority calculation -> I am using edge differences instead of the edge quotients mentioned in the paper, which seem to perform worse. I have not spent too much time trying different linear combinations of the different terms, but whenever I tried varying them it seemed like not too much is changing. --> I suspect this can be used to further improve performance, but this probably is not the main issue.

3. Try different witness path search algorithms as implemented in `EdgeBasedNodeContractor`. The turn replacement algorithm should already give the least amount of shortcuts, but the `SimpleSearch` is simpler and takes less effort. However, right now I think both variants introduce 'too many' shortcuts and take too long to prepare the graph. I was also thinking about a hybrid where we do the 'costly' turn replacement search or the 'fast' simple search depending on some metric, but right now I do not think this is the bottleneck.

4. Collect all shortcuts to be inserted for a given node and do a batch insert instead of adding them one by one (this is currently done in the node-based version of CH) --> did not seem to help much, because I am checking for duplicate shortcuts when the shortcuts are inserted anyway.

5. One thing I am doing different than in the paper is that for loop shortcuts I recursively add some helper shortcuts to be able to unpack them later. However, I found that the number of these helper shortcuts is rather small and I (no longer) think that this is the root of the problem of so many shortcuts.

6. Limiting of local searches: We can not cancel the search entirely, because the witness path search needs to 'find' the shortcut connection (via loops). However, we can discontinue all active leafs of the search that do not belong to a possible connection consisting only of the in- and out edges that are currently checked and loops at the node to be contracted. This is done when the number of settled edges exceeds a certain number (not given in the paper). If this number is too low some witnesses will not be found, but even increasing the limit to a very high number did not significantly reduce the number of shortcuts. So again, I think there is room for some improvement by tuning this parameter, but I could not get a significant improvement doing so so far.

7. Using fwd/bwd flags on shortcut edges. Right now all shortcuts are used with only the fwd flag. If we use fwd/bwd flags we can potentially reduce the number of EdgeIteratorStates by 50% , which should be beneficial, but it requires some changes, because we then need a more involved way to decide what the 'first' and 'last' original edges of a shortcut are (similar to the edge-key calculation where the two node ids of an edge are used to decide if it is fwd or bwd). So this would definitely improve performance I think, but it does not reduce the 'real' number of shortcuts.

8. One interesting thing I found was in https://github.com/graphhopper/graphhopper/pull/1247/commits/382432d5a09f9613bee6a8a9ac5f52fd7647e499 where I disabled a previously introduced simple check for duplicate neighbor updates. Disabling this check improved the query time by a factor of 13 (!) on the map for Bremen without changing the number of shortcuts very much (less than 1% difference). I currently have no understanding why this makes such a large difference.

It could be that the rather poor performance can only be improved by a combination of all of the above things, but I am still hoping there is some kind of bottleneck that can be fixed or I am simply forgetting something important. 

@karussell (or anybody else interested) maybe you could have a look and notice something ? Not sure how familiar you are with the turncost version of CH, but maybe there is something that was crucial for node-based CH that I am missing ?

For testing I am currently using the CHMeasurement class, which performs the contraction for given OSM data and compares some random queries with normal Dijkstra. It usually also logs some errors, where the calculated weights are different, but they should be very small (I think some rounding issue that I did not find the reason for yet). Afterwards the query time is also measured. During the contraction some statistics are logged (for example we see that a lot of time is spent on neighbor updates and the number of shortcuts that are checked rises dramatically at the end).

karussell commented:
Thanks!

I have no idea for a solution but I will have a look and here are my 2 cents:

 * utilizing a shortcut for fwd and bwd give a good boost in my implementation but it makes everything much more complicated and was one of the bigger bug pools :(
 * tuning the heuristic (and some bug fixing, see point before) was a large part of the time I invested
 * If especially the last few percentages introduce so many shortcuts there was a similar problem where the priority recalculation took too long at the end and so I allowed this only for `sortedNodes.getSize() < lastNodesLazyUpdates`. Stopping the contraction should be completely ok (see nodesToAvoidContract), but for the node based CH the benefits of saved time and storage in the preparation did not outweigh the performance losts. It was like e.g. already when I did the contraction for 99.9% there was already a measurable slow down at query time.
 * In my cases it helped a bit doing the contraction for bigger areas like Germany or if possible Europe to get an idea which shortcuts are unnecessary or similar. But debugging this stuff is hard and takes long. E.g. It took 3 months from having Germany running on my laptop until world wide CH routing was properly working, initially with preparation times of more than 4 days.

Maybe we can at some point do a brainstorming call :) ?

karussell commented:
> I currently have no understanding why this makes such a large difference.

This is interesting and might be important to understand. Especially in this case - maybe the heuristic for node-based-CH needs other things than the edge-based variant.

ammagamma commented:
> utilizing a shortcut for fwd and bwd give a good boost in my implementation but it makes everything much more complicated and was one of the bigger bug pools :(

Yes, this is my fear also and the reason why I postponed it so far. However, I think sooner or later it will be required.

> tuning the heuristic (and some bug fixing, see point before) was a large part of the time I invested

Just knowing that this took some time already helps and makes me feel less stupid ;)

> Maybe we can at some point do a brainstorming call :) 

Yes, this would be nice! I will do a bit more analysis, but then I think this would be a very good idea.

> If removing this really makes the query faster then I think those neighbor updates are not 'real' duplicates. Also it might be worth to play with neighborUpdatePercentage.

This duplicate neighbor check did not exist in the CH algorithm so far, I introduced it in hope of making the contraction faster because I did not see any reason to update a neighbor node multiple times just because there are multiple edges leading to it (which can happen quickly in edge-based CH). It definitely slows down the queries so I removed it / commented it out for now. The most surprising thing for me is that the number of shortcuts is not really affected, but the query time changes so much. I agree that we should check carefully how the neighbor updates affect the performance, and I have not done this systematically yet.

ammagamma commented:
As it is often with 'very interesting' behavior it turns out the slow down of the routing queries when the duplicate neighbor update check was enabled was simply due to a stupid error: The check not only prevented duplicate neighbor updates but also important edge disconnects. Sorry for the confusion.

karussell commented:
> Sorry for the confusion

Really no problem and good that you sorted out (this smaller issue :))

ammagamma commented:
Luckily I was wrong with a few things. The local search limiting had an error and I also made it dynamic now. Also I improved the turn replacement search such that it finds more witnesses, but then I changed the default to what is described as 'aggressive' search in the paper, which does more computation but also introduces less shortcuts. The aggressive search also benefits from an array-based version of the WitnessPathFinder that I added now. My current 'record' is the contraction of the map of Germany in about 2.5h with query times around 100ms. For Berlin the contraction time is now down to 90s and the queries are 2x faster now (around 7ms).

Right now there are some deviations from Dijkstra queries that are due to duplicate edges in the OSM data, and I am still trying to find a better way to handle them.

karussell commented:
Just a minor question: why is the int-int TreeMap moved to float-float?

ammagamma commented:
> Just a minor question: why is the int-int TreeMap moved to float-float?
Do you mean GHTreeMapComposed ? It used to store two ints in the Long key, now it is an int and a float. The float is used for the node priority of the node with the id that is stored in the int. For edge based CH it turned out  that edge quotients work much better than edge differences so a float is more useful for this. The tree map was not used for anything else than Contraction Hierarchies, so I changed both node- and edge- based CH to use floats for node priorities. Of course we could create a new class instead and keep the old GHTreeMapComposed with two ints. I also tried to store the quotient as an int by multiplying it with some fixed 'precision', but this did not seem to work (and I turned back to edge differences back then...)

karussell commented:
Ah, thanks for clarification.

karussell commented:
I think we can remove "WIP" from the title already?

Could we even already merge this and improve with further & smaller PRs?

The only minor thing we should implement is a deactivation this if people want to avoid the edge-based CH preparation overhead. We probably will never get too close to the node based version and turn costs is sometimes not required for CH but still for non-CH.

Potential follow up PRs: 1. reduce the traversal modes to just the node and edge based. 2. faster request 3. faster import etc

ammagamma commented:
> I think we can remove "WIP" from the title already?
Could we even already merge this and improve with further & smaller PRs?

I'm glad you think this is getting close to being merge-able. I also think it might make sense to continue with smaller PRs mostly because this should make the review easier to follow. I think its best to even split this current PR into smaller ones, because some stuff does not only affect edge-based CH (for example the GenericBidirectionalDijkstra refactoring) and with a smaller PR it would be easier to detect any possible regressions.
At the same time there is some stuff that I am not sure if it is worth merging into master, specifically I am not fully settled on some details of the node contraction, yet. Also there are still some cases where there are some deviations from normal Dijkstra and I have not yet found the reason for this.

 > The only minor thing we should implement is a deactivation this if people want to avoid the edge-based CH preparation overhead. We probably will never get too close to the node based version and turn costs is sometimes not required for CH but still for non-CH.

I agree, but not sure how this should be done. Right now edge-based CH will be enabled when using e.g. `car|turn_costs=true` for `graph.flag_encoders` and CH will be disabled by using `prepare.ch.weightings=no`. So either there will be no CH at all or no turn costs. I think we need another property for this ? Also I was wondering if we should make it possible to configure node-based CH and edge-based CH at the same time such that we can switch between turn-cost CH (slower) and node-based CH (faster but no turn restrictions) at query time ? See `GraphHopperIT#testTurnCostsOnOffCH` and `GraphHopperIT#testCHOnOffWithTurnCosts`.

devemux86 commented:
> So either there will be no CH at all or no turn costs.

Indeed it's essential the flexible mode to continue working without lesser features, as it offers much flexibility than always needed turn costs, like headings, alternatives, etc.

ammagamma commented:
And btw +1 for the 'algorithm' label :)

karussell commented:
Yes, we need a new property that should even be per encoder, so maybe something clear but lengthy like `car|turn_costs=true|ch.turn_costs=true`?

> At the same time there is some stuff that I am not sure if it is worth merging into master

Sure, please take the necessary time, no need to hurry and of course (as you already did) we should pick stuff from this PR into a separate if possible.

> Also I was wondering if we should make it possible to configure node-based CH and edge-based CH at the same time such that we can switch between turn-cost CH (slower) and node-based CH (faster but no turn restrictions) at query time ?

Cool. Didn't know that this is possible. What are the percentage differences in terms of speed?

ammagamma commented:
> Cool. Didn't know that this is possible. What are the percentage differences in terms of speed?

I just meant if you have the memory you could prepare and load the node-and edge-based CH graphs on the server and then use whatever is requested in the query, but I did not do anything in this direction yet (except writing the failing test :)

ammagamma commented:
> Yes, we need a new property that should even be per encoder, so maybe something clear but lengthy like `car|turn_costs=true|ch.turn_costs=true`?

I am hesitant to put this in the encoder string, because the ch preparations do not have anything to do with encoders and also doing it this way does not allow to trigger both node- and edge-based CH preparation for the same vehicle.  

With the current syntax `car|turn_costs=true` we only make sure that the `maxTurnCosts` for the car encoder are set to 1 (same for some other encoders) and as soon as there is at least one vehicle with `turn_costs=true` there will be a `TurnCostExtension` for the graph. The other thing is which CH preparations should be run for the given vehicles/encoders. At the moment we only can choose between `ch.prepare.weightings=no` and `ch.prepare.weightings=<one or more weightings>`, so either no CH preparation is done at all or it is done for all vehicles (and for all given weightings). If we want to make this completely configurable we would need a way to decide for each vehicle if node- and/or edge-based CH should be prepared or not. Both have benefits and I do not think that edge-based CH will be preferable if one does not need turn costs and performance is more important.
One easy (but not fully flexible) way would be to add a property `prepare.ch.turn_costs` with three different options to choose from:

- `no` -> no edge-based CH, like it is in master now
- `additional` -> edge-based CH will be prepared for all encoders with turn_costs=true and all weightings in `ch.prepare.weightings` additionally to the node-based CH preparations
- `only` -> like `additional` but no node-based CH will be prepared for encoders without turn_costs 

Would this be sufficient for now ?

I think to be most flexible we would need a list of vehicle profiles like this: 
```yaml
[
vehicle: car
encoderOptions: turn_costs=true,speed_factor=3
ch.weightings.node_based: fastest,shortest
ch.weightings.edge_based: fastest
lm.weightings:shortest
# maybe more preparations in the future here...
,
vehicle: bike
...
]
```
but I am sure this needs some further thought and would be rather problematic in terms of backwards compatibility.

karussell commented:
> Would this be sufficient for now ?

Yes!

Maybe we make the intent more clear for the `additional` case, like with `node_and_edge_based`? And then for `only` we use `edge_based` and/or just `yes` or `true`?

> I think to be most flexible we would need a list of vehicle profiles like this
> would be rather problematic in terms of backwards compatibility

As long as it is easy to convert the yml to this new format the backward compatibility is not a big issue, this should be probably discussed in #493. (btw: the web API is the only thing that we cannot experiment much and where we would take special care with e.g. a separate endpoint)

ammagamma commented:
ok! and thank for the pointer to #493 I must have missed this.

develar commented:
@ammagamma Great that you continue work on this feature :)

easbar commented:
I think this is ready for review.

karussell commented on core/src/main/java/com/graphhopper/routing/ch/ManualPrepareContractionHierarchies.java:
```diff
@@ -0,0 +1,123 @@
+        for (int i = 0; i < nodes; ++i) {
+            result.add(i);
+        }
+        // the shuffle method is the only reason we are using java.util.ArrayList instead of hppc IntArrayList here
```
:) we should avoid ArrayList as it uses much more space for large node count and is probably also slower. (Here we could even use an int array)

easbar replied:
I agree, but I did not know of any easy way to shuffle an int array ? Right now this code is not used anywhere anyway and I removed it for now. This whole class will be reworked/moved when doing #1509.

karussell commented on core/src/main/java/com/graphhopper/storage/GraphHopperStorage.java:
```diff
@@ -420,4 +452,14 @@ public Graph copyTo(Graph g) {
+        // the lists are in sync). another option would be something like guava concat.
+        List<CHGraphImpl> result = new ArrayList<>();
+        result.addAll(nodeBasedCHGraphs);
+        result.addAll(edgeBasedCHGraphs);
```
A minor optimization here is to create the new ArrayList with the nodeBasedCHGraphs already or at least init the size with node+edge

easbar replied:
I did not want to create the list already, because then one has to make sure that it is in sync with the other two. Right now this is easy because the lists are changed only in the constructor, but I was afraid might change at some point and thought it is not worth it. I do not think this method is never called in a performance-relevant code. I added the capacity parameter to the ArrayList constructor for now.

easbar commented:
> Will have to study it a bit more in the next days, especially the WitnessPathSearcher. Therefore have only unrelated and not really helpful comments ;)

Thanks for taking a look and take your time ;)

> BTW: Should we remove all the other edge based traversals and just keep node based and EDGE_BASED_2DIR? (in a different PR)

I think so, yes.

easbar commented on core/src/main/java/com/graphhopper/routing/AStarBidirectionEdgeCHNoSOD.java:
```diff
@@ -0,0 +1,112 @@
+        ((AStarCHEntry) entry).incEdge = edgeId;
+        entry.weight = getHeapWeight(edge.getAdjNode(), reverse, weight);
+        ((AStarCHEntry) entry).weightOfVisitedPath = weight;
+        entry.parent = parent;
```
I think we had the generic parameter further up in the hierarchy in AbstractBidirAlgo before, but then removed it again. Hmm, I think I would leave this as it is for now, but I do not really care either.

easbar commented on core/src/main/java/com/graphhopper/routing/ch/EdgeBasedNodeContractor.java:
```diff
@@ -0,0 +1,550 @@
+        // we need to set flags first because they overwrite weight etc
+        shortcut.setFlags(direction);
+        shortcut.setSkippedEdges(edgeFrom.edge, edgeTo.edge)
+                // this is a bit of a hack, we misuse incEdge of the root entry to store the first orig edge
```
yes, I mean edgeFrom's parent, fixed.

easbar commented on core/src/main/java/com/graphhopper/routing/ch/EdgeBasedNodeContractor.java:
```diff
@@ -0,0 +1,550 @@
+
+        @Override
+        public void findAndHandleShortcuts(int node) {
+            LOGGER.debug("Finding shortcuts (aggressive) for node {}, required shortcuts will be {}ed", node, activeShortcutHandler.getAction());
```
Using debug and trace did not really make sense here, because both logs would very frequent when they are enabled. I used this mostly for debugging. I set all to trace now.

easbar commented on core/src/main/java/com/graphhopper/util/CHEdgeIteratorState.java:
```diff
@@ -36,7 +36,16 @@
+     * <p>
+     * x --> u --> v --> w --> y
+     */
+    CHEdgeIteratorState setOuterOrigEdges(int firstOrigEdge, int lastOrigEdge);
```
Yes, I agree and have renamed it to setFirstAndLastOrigEdges and the getters as well.

easbar commented on core/src/main/java/com/graphhopper/routing/ch/CHEntry.java:
```diff
@@ -0,0 +1,43 @@
+import com.graphhopper.util.EdgeIterator;
+
+public class CHEntry extends SPTEntry {
+    public int incEdge;
```
I added some docs.

karussell replied:
Thanks! There is an `is` missing: "but for shortcuts this the id ..."

easbar commented on core/src/main/java/com/graphhopper/routing/AbstractBidirAlgo.java:
```diff
@@ -67,7 +67,7 @@ protected void initCollections(int size) {
     protected abstract SPTEntry createStartEntry(int node, double weight, boolean reverse);
 
-    protected abstract SPTEntry createEntry(EdgeIteratorState edge, double weight, SPTEntry parent, boolean reverse);
+    protected abstract SPTEntry createEntry(EdgeIteratorState edge, int edgeId, double weight, SPTEntry parent, boolean reverse);
```
Ok I have added java-docs, do the help with the understanding ? I have also renamed the edgeId parameter to incEdge.

easbar commented on core/src/main/java/com/graphhopper/routing/ch/CHAlgoFactoryDecorator.java:
```diff
@@ -130,24 +151,42 @@ public CHAlgoFactoryDecorator addWeighting(String weighting) {
-        int lastIndex = preparations.size() - 1;
-        if (lastIndex >= weightings.size())
+        // we want to make sure that edge- and node-based preparations are added in the same order as their corresponding
+        // weightings, but changing the order between edge- and node-based preparations is accepted
```
Hmm, I know this API is a bit tricky ;) Not sure if it really becomes much better when we enforce having to add the node based preparations before the edge based ones ? If we did it like this we would get rid of calculating the index above, but would have to store the end of the last index of a node-based preparation instead (and throw an error if a node preparation is added after the first edge preparation). If you prefer this option I'll do it otherwise I would leave as is :)

easbar commented:
> The performance is identical to master for my tests. Nice!

Yes, this is what I have found, too.

> PS: there seems to be some minor problems with the forbidden API check regarding the String.format call
I fixed one String.format where the default locale was used (I think in GHPoint). However the tests still sometimes fail because of OVERNIGHT even during (European) daytime ?

easbar commented:
> What do you mean here?

I mean this here: https://github.com/graphhopper/graphhopper/pull/1247/commits/8be07a1fc9bbe7c3a6e542ff6f3c2ee2dc38cac4 Without specifying the locale appveyor fails.

> BTW: Often just a matter of taste but for many parameters I find String.format more complicated (due to the associations forth&back) than string concatenating.

I know :) I often like String.format better (for example when floats are involved and I want to restrict to a few decimals like in the case here). But please always feel free to comment when you want to change it. I think both methods have their pros and cons and we should decide from case to case ?

easbar commented:
> What I can see in the logs is:

Yes this is the same issue, will fix.

karussell commented:
> Without specifying the locale appveyor fails.

The reason is not appveyor. It is more the forbidden api plugin that we call which enforces to avoid certain API that use unintended defaults, so it forces us to be more specific. You can do so also locally: https://github.com/graphhopper/graphhopper/blob/master/.travis.yml#L42

karussell commented:
> I know :) I often like String.format better (for example when floats are involved and I want to restrict to a few decimals like in the case here).

Good point - never thought about this. So surely we should decide from case to case.

Now I can see what you mean with OVERNIGHT ... have updated master in the hope that those problems are fixed

> [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.519 s <<< FAILURE! - in com.graphhopper.AnotherAgencyIT
> [ERROR] testRoute1(com.graphhopper.AnotherAgencyIT)  Time elapsed: 0.038 s  <<< ERROR!
> java.lang.RuntimeException: OVERNIGHT
>	at com.graphhopper.AnotherAgencyIT.testRoute1(AnotherAgencyIT.java:83)

easbar commented:
> Now I can see what you mean with OVERNIGHT ... have updated master in the hope that those problems are fixed

Did you push these changes ? I think I have fixed all forbidden apis, but some of the integration tests still fail with the OVERNIGHT error.

karussell commented:
This OVERNIGHT RuntimeException is something different, I think. As we merged master (where it does not occur). @michaz do you have a clue what this could be or mean?

michaz commented:
That's serious, I'm afraid.. It basically means that something interferes with "my" custom attributes that I put into the edge flags.

easbar commented:
> That's serious, I'm afraid.. It basically means that something interferes with "my" custom attributes that I put into the edge flags. See PtFlagEncoder.
Maybe some assumptions about how to combine flag encoders changed?

Oh, I always thought this is because we cannot run the integration tests that make requests against GH apis at night :D What does OVERNIGHT stand for ? There is basically nothing in this branch related to flags or encoders, but I will have a look.

michaz commented:
> OVERNIGHT

Never mind that word -- it's just that in this particular case an assertion is violated, the router finds an edge type it didn't expect. Edge type is one of the attributes I store.

When I remove that particular assert, other things go wrong instead, which point to the same problem.

michaz commented:
Other ideas: Maybe, in the PT execution path, something happens now that didn't happen before? Some preprocessing, ... _something_ that the PT router doesn't expect and didn't ask for?

karussell commented:
If I uncomment "if (iter.getBaseNode() == iter.getAdjNode()) {" in DefaultEdgeFilter the PT tests are green

karussell commented:
Sorry. I meant comment (not uncomment)

karussell commented:
I remember in PT we use some special edges that have a node pointing to itself to store information on a node. Is this correct? Why would it be wrong to always return true for the edge filter? See the code here:

https://github.com/easbar/graphhopper/blob/turncosts_ch/core/src/main/java/com/graphhopper/routing/util/DefaultEdgeFilter.java#L54-L63

easbar commented:
> If I uncomment "if (iter.getBaseNode() == iter.getAdjNode()) {" in DefaultEdgeFilter the PT tests are green

Yes, this is what I just found, too ;)

easbar commented:
If this is a problem I think we could use a dedicated filter for edge-based CH, but if we can avoid "using some special edges that have  a node pointing to itself to store information on a node" I think this would even be better because such workarounds sometimes quickly can become problematic for edge-based algorithms (e.g. #1355 )

michaz commented:
Ah, I see. Yeah, I'm using the edge filter to only give me edges that the ``foot`` flag encoder accepts (i.e. only the road network). And that fails now. It now also gives me some internal PT loop edges that I never expected to start a trip on.

> Why would it be wrong to always return true for the edge filter?

Because then it wouldn't filter anything?

Couldn't this also lead to problems in other cases? If I'm only looking for ``bike`` edges to start a trip on, I only want ``bike`` edges, and not some loops that may be on a different network..

michaz commented:
(I could _probably_ do without those loops. But somehow "I'm going to filter edges, _except_ loops, which I will always return no matter what" seems a weird thing for a ``DefaultEdgeFilter`` to do, I think.

karussell commented:
Yes, I think you are right. I guess real loops "A-A" in road networks are rare in practise but they exist:

https://graphhopper.com/maps/?point=51.438751%2C14.232708&point=51.439524%2C14.232649

karussell commented:
Let's assume that one vehicle profile has access to this loop (service road) and one like e.g. a big truck has not. Would this mean that currently we would still give it access to the road @easbar ?

easbar commented:
DefaultEdgeFilter looks for the direction flags and either accepts fwd (where iter.isForward() = true), bwd (where iter.isBackward()=true) or edges that are both fwd and bwd. This is also reflected in the static factory methods outEdges, inEdges, allEdges. For a loop the meaning of the fwd and bwd flags is not well defined, it cannot be pointing away or towards the node, it always does both. 

> Let's assume that one vehicle profile has access to this loop and one like e.g. a big truck has not. Would this mean that currently we would still give it access to the road @easbar ?

How would the vehicle profile restrict the access to the loop ? The filter only checks the direction flags of the edge. I do not think their will be an encoder that always makes iter.isForward/Backward(encoder) return true/false ? 

> But somehow "I'm going to filter edges, except loops, which I will always return no matter what" seems like a weird thing for a DefaultEdgeFilter to do, I think.

If our options are outEdges, inEdges and allEdges I think it somewhat makes sense that loops are always accepted because they fit all these categories. 

Maybe it is hard to agree on a 'default' edge filter that works for all use cases and probably there is not really a need to share this filter across such different domains as PT and CH ? However DefaultEdgeFilter is used pretty high in the routing algorithm hierarchy so it might make sense to use the current implementation as DefaultRoutingEdgeFilter ?

easbar commented:
Otherwise we could also solve this 'problem' by using special EdgeFilters (with if(base=adj) return) in the edge-based CH algos (ignoring the ones inherited from AbstractRoutingAlgorithm etc.)

easbar commented:
Also I think a better name for `DefaultEdgeFilter` would be `EdgeDirectionFilter` :)

michaz commented:
> The filter only checks the direction flags of the edge.

.. which are _per profile_. That's what the parameter ``encoder`` to ``iter.isForward(..)`` means. Not super-obvious.

I'm using this filter like "is this an edge (either a forward edge or a backward edge) _for walking_". And now you always say "yes", no matter for which profile. That's the only problem.

michaz commented:
How about replacing ``true`` with 

```
iter.isForward(encoder) || iter.isBackward(encoder)
```

Would that do it for you?

easbar commented:
> Would that do it for you?

Yes, I think so. This way you can still set isForward/isBackward(encoder) to false in a profile to filter out the edge. Sorry, I was not aware this would make sense for any profile. I thought you would rather set the distance to infinity to block an edge ?

michaz commented:
Okay, cool. It still seems _kind of_ dangerous / special-case-ish.

I changed my mental model of our graph several times, but I currently think it is an _undirected graph_, where edges have _one set of attributes per direction_, and one family of those attributes-per-direction happens to be "is accessible for mode X".

Now with loops, the problem is that they don't have a "direction" in the graph-theoretical sense, but the word "direction" above is not meant in the graph-theoretical sense: In the data structure, even the loop has two directions, hence two sets of attributes.

Now when I filter for edges that go in "one direction", I will still get loops even if they go in the "other direction".

michaz commented:
(Talking to myself now) On the other hand, this seems to be an argument to _forbid_ loops that have "both directions" accessible. Because that would mean that the transformation to a directed graph would  _duplicate_ loops (and only loops), which would make it a multigraph and not a graph.

easbar commented:
Thanks for your explanations!
Unfortunately there is another test that fails: `IsochroneResourceTest#requestReverseFlow`. This one involves a loop with isForward=true and isBackward=false. I do not understand what this is supposed to mean (it could just as well be the other way around, without any real meaning), but apparently this loop needs to be rejected by the filter to make the test pass...
To make things even more complicated compared to normal loop edges I think the shortcuts loops in edge-based CH do have a meaningful direction which is defined by their first/last original edges. This is the reason I needed the check in `DefaultEdgeFilter` in the first place. 

>  On the other hand, this seems to be an argument to forbid loops that have "both directions" accessible

I just thought the opposite. I would forbid loops that only have one direction accessible, because there is no way to differentiate between the two (besides edge-based shortcuts, see above...). Or maybe I would just forbid any kind of loops... ;) But I do get your point about the transformation to a directed graph, this is also true :)

easbar commented:
To give an example, have a look here: https://github.com/easbar/graphhopper/blob/2375b34b211b379c319ba3bb14966462cb7d0498/core/src/test/java/com/graphhopper/routing/ch/EdgeBasedNodeContractorTest.java#L1139

There is a (fwd-only) loop at node 1 and we need to determine the edges incoming to this node. `DefaultEdgeFilter.inEdges` will not find it, because the loop returns isBackward=false. When instead we would  use `DefaultEdgeFilter.outEdges` we would find the loop, but there is no real reason why it should not be the other way around or it should even make a difference which filter we are using. All we did was call `graph.edge(1, 1, dist, false)` and this (quite arbitrarily) sets the fwd, but not the bwd flag. I think from this point of view `graph.edge()` should just have set both (fwd&bwd) flags.

karussell commented:
> I would forbid loops that only have one direction accessible, because there is no way to differentiate between the two 

The geometry is one thing which can be used to differentiate between the two. Assume a one-way loop for cars that allows both directions for foot and is accessible only in the reverse direction for bike - such a complicated case should be possible already from the routing and storing side. I remember having issues with one-way loops see #162.

@easbar in the (mergable) edge flags branch we have the flags always in the same direction - would this somehow help?

karussell commented:
Likely the Isochrone test needs a fix as it needs access to the one-way loop too regardless of the edge filter, just like in your edge-based case.

The node-based routing is different: either, when there is no waypoint on a loop, we do not care about the loop or we create a virtual node via QueryGraph to e.g. start routing there which avoids this uncertainty.

easbar commented:
> The geometry is one thing which can be used to differentiate between the two

Ok, yes. This also fits with the way of thinking @michaz explained. We cannot differentiate between the two the way the routing is done at the moment, but this is for another issue. So we need a way to have loops that are only forward or backward.

easbar commented:
> From my side we can merge it.

yay!

> Or do you know some git magic to reduce the number a bit :) ?

Hmm the only thing would be interactive rebase with some squashes ? There are quite a few 'fixes/minor/less important' commits so maybe it would be reduced to about 150 ? Then again sometimes when there are merge commits in between one has to resolve the conflicts again, which could be annoying. In any case this way  we would still lose the git hashes (probably sometimes used in the discussion here). Either way I will keep this branch and I can also add it to the core repo. I think I would have squashed everything into one commit, but whatever you like better.

> Also you need to store the Email you used in the commits into your github account (I think you can easily add it even if you do no longer use actively it), otherwise those many commits do not belong to you.

Ok, thanks for pointing this out, I added my mail&pgp key.

I added some `@author` tags btw, hope this is fine.
